File: go-sdk/pkg/observability/tracing/propagation.go
package tracing

import (
	"context"
	"net/http"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/trace"
)

// ExtractContext extracts trace context from HTTP headers
func ExtractContext(r *http.Request) context.Context {
	return otel.GetTextMapPropagator().Extract(r.Context(), propagation.HeaderCarrier(r.Header))
}

// SpanFromContext returns the current span from context
func SpanFromContext(ctx context.Context) trace.Span {
	return trace.SpanFromContext(ctx)
}

// ContextWithSpan creates a new context with the given span
func ContextWithSpan(ctx context.Context, span trace.Span) context.Context {
	return trace.ContextWithSpan(ctx, span)
}

// GetTraceIDFromContext extracts trace ID from context
func GetTraceIDFromContext(ctx context.Context) string {
	spanCtx := trace.SpanContextFromContext(ctx)
	if spanCtx.HasTraceID() {
		return spanCtx.TraceID().String()
	}
	return ""
}

// GetSpanIDFromContext extracts span ID from context
func GetSpanIDFromContext(ctx context.Context) string {
	spanCtx := trace.SpanContextFromContext(ctx)
	if spanCtx.HasSpanID() {
		return spanCtx.SpanID().String()
	}
	return ""
}

File: js-sdk/src/observability/tracing/propagation.ts
/**
 * Trace context propagation and span access.
 * Aligns with Go ExtractContext, SpanFromContext, GetTraceIDFromContext, etc.
 */

import { context, trace, INVALID_SPAN_CONTEXT } from '@opentelemetry/api';
import type { Span } from '@opentelemetry/api';

/**
 * Get the current span from the active context.
 */
export function spanFromContext(ctx: ReturnType<typeof context.active>): Span {
  return trace.getSpan(ctx) ?? trace.wrapSpanContext(INVALID_SPAN_CONTEXT);
}

/**
 * Get the current trace ID from the active span context.
 * Returns hex string or empty if no valid span.
 */
export function getTraceId(): string {
  const span = trace.getActiveSpan();
  if (!span) return '';
  const spanContext = span.spanContext();
  if (!spanContext.traceId) return '';
  return spanContext.traceId;
}

/**
 * Get trace ID from a given context.
 */
export function getTraceIdFromContext(ctx: ReturnType<typeof context.active>): string {
  const span = trace.getSpan(ctx);
  if (!span) return '';
  const spanContext = span.spanContext();
  return spanContext.traceId ?? '';
}

/**
 * Get span ID from a given context.
 */
export function getSpanIdFromContext(ctx: ReturnType<typeof context.active>): string {
  const span = trace.getSpan(ctx);
  if (!span) return '';
  const spanContext = span.spanContext();
  return spanContext.spanId ?? '';
}

export { context };
